main/src/library/Fixpoint3/Interpreter.flix

```diff
@@ -0,0 +1,699 @@
+                result |> Vector.forEach(t -> {
+                    writes |> Vector.forEach(match (srcIndex, dst, dstIndex) -> {
+                        Array.put(Vector.get(srcIndex, t), dstIndex, Vector.get(dst, minEnv));
+                        Array.put(Vector.get(srcIndex, t), dstIndex, Vector.get(dst, maxEnv))
+                    });
+                    Array.put(t, tupleIndex, tupleEnv);
+                    evalOp(rc, ctxt, parLevel, body)
+                })
+        }
+
+    ///
+    /// Handle all writes in `writes` reading from `t` and writing to `minEnv` and `maxEnv`.
+    ///
+    /// For `t=[8, 42]`, `writes=[(1, 2, 0)]` it would do the update `minEnv[2][0] = t[1]`.
+    /// The same update is done for `maxEnv`.
+    ///
+    /// Lastly `t` is saved at `tuplePos` in `tupleEnv`, i.e. `tupleEnv[tuplePos] = t`.
+    ///
+    def handleWriteTuples(
+        t: Tuple,
+        writes: Vector[WriteTuple],
+        minEnv: Vector[SearchTuple[r]],
+        maxEnv: Vector[SearchTuple[r]],
+        tuplePos: Int32,
+        tupleEnv: Array[Tuple, r]
+    ): Unit \ r =
+        writes |> Vector.forEach(match (srcIndex, dst, dstIndex) -> {
+            Array.put(Vector.get(srcIndex, t), dstIndex, Vector.get(dst, minEnv));
+            Array.put(Vector.get(srcIndex, t), dstIndex, Vector.get(dst, maxEnv))
+        });
+        Array.put(t, tuplePos, tupleEnv)
+
+    ///
+    /// Handles executing `body` for a `Search(Relational)` or `Query(Relation)`
+    /// RelOp with current bound tuple `t`.
+    ///
+    /// Should only be used when new threads are being spawned.
+    ///
+    /// `body` is the body of the `Search` or `Query`. `writes` is a vector of writes to `SearchTuple`'s
+    /// to be performed. See `handleWriteTuples`.
+    ///
+    def parRelationHandlingBodyFunc(
+        rc: Region[r],
+        tuplePos: Int32,
+        writes: Vector[WriteTuple],
+        body: RelOp,
+        ctxt: Context[r],
+        parLevel: AllowedBranches,
+        t: Tuple,
+        ignored: Boxed
+    ): Unit \ r =
+        relationHandlingBodyFunc(rc, tuplePos, writes, body, makeCtxt(rc, ctxt), parLevel, t, ignored)
+
+    ///
+    /// Clones `ctxt` for a concurrent thread.
+    ///
+    def makeCtxt(rc: Region[r], ctxt: Context[r]): Context[r] \ r = {
+        let (db, (minEnv, maxEnv, tupleEnv, latEnv), boxing) = ctxt;
+        let threadMinEnv = minEnv |> Vector.map(x -> Array.copyOfRange(rc, 0, Array.length(x), x));
+        let threadMaxEnv = maxEnv |> Vector.map(x -> Array.copyOfRange(rc, 0, Array.length(x), x));
+        let threadTupleEnv = Array.copyOfRange(rc, 0, Array.length(tupleEnv), tupleEnv);
+        let threadLatEnv = Array.copyOfRange(rc, 0, Array.length(latEnv), latEnv);
+        let threadEnv = (threadMinEnv, threadMaxEnv, threadTupleEnv, threadLatEnv);
+        (db, threadEnv, boxing)
+    }
+
+    ///
+    /// Handles executing `body` for a `Search` or `Query` relation with current bound tuple `t`.
```

Comment:
Instead of these "handles blah". Tell me what it does :)

Reply:
"Handles insertion into a red blacktree" does not say much. But bringing words up like binary search, rotations, recoloring etc helps. Similar comments here and elsewhere would be useful.